source/air/src/printer.rs

```diff
@@ -495,3 +495,166 @@ impl NodeWriter {
 pub(crate) fn node_to_string(node: &Node) -> String {
     NodeWriter::new().node_to_string_indent(&"".to_string(), node)
 }
+
+pub fn singular_query_to_node(vars: Vec<String>, enss: Vec<Expr>, reqs: Vec<Expr>) -> Node {
+    let pp = Printer::new(false);
+    let mut nodes = Vec::new();
+    let mut var_nodes = Vec::new();
+    nodes.push(str_to_node("singular-check-valid"));
+    let ens = pp.exprs_to_node(&Arc::new(enss));
+    let req = pp.exprs_to_node(&Arc::new(reqs));
+    for v in vars {
+        var_nodes.push(Node::Atom(v));
+    }
+    nodes.push(Node::List(var_nodes));
+    nodes.push(ens);
+    nodes.push(req);
+    Node::List(nodes)
+}
+
+use std::collections::HashMap;
+pub(crate) fn expr_to_singular(
+    expr: &Expr,
+    tmp_idx: &mut u32,
+    fmap: &mut HashMap<Node, String>,
+) -> String {
+    match &**expr {
+        ExprX::Const(Constant::Nat(n)) => n.to_string(),
+        ExprX::Var(x) => x.to_string(),
+        ExprX::Binary(BinaryOp::EuclideanMod, lhs, rhs) => {
+            // x % y ->  x - y*tmp
```

Comment:
might want to use consistent temp variables for the same x and y

Reply:
You're right, thank you!